plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/editors/sql/semantics/model/SQLQueryRowsCteModel.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * DBeaver - Universal Database Manager
+ * Copyright (C) 2010-2024 DBeaver Corp and others
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jkiss.dbeaver.ui.editors.sql.semantics.model;
+
+
+import org.antlr.v4.runtime.misc.Interval;
+import org.jkiss.code.NotNull;
+import org.jkiss.code.Nullable;
+import org.jkiss.dbeaver.ui.editors.sql.semantics.SQLQueryRecognitionContext;
+import org.jkiss.dbeaver.ui.editors.sql.semantics.SQLQuerySymbolClass;
+import org.jkiss.dbeaver.ui.editors.sql.semantics.SQLQuerySymbolEntry;
+import org.jkiss.dbeaver.ui.editors.sql.semantics.context.SQLQueryDataContext;
+
+import java.util.ArrayList;
+import java.util.List;
+
+
+public class SQLQueryRowsCteModel extends SQLQueryRowsSourceModel {
+
+    private final boolean isRecursive;
+    private final List<SQLQueryRowsCteSubqueryModel> subqueries = new ArrayList<>();
+    private final SQLQueryRowsSourceModel resultQuery;
+
+    public static class SQLQueryRowsCteSubqueryModel extends SQLQueryRowsSourceModel {
```

Comment:
Why is it nested?

Reply:
The idea is that it shouldn't exist without containing CTE model.
But it's public, because we want to see it in Outline, for example.
Do you think it's better to separate them?

Reply:
I see. It's just my personal taste, but I see reasoning behind this decision, so it's ok